Networking

richard1 opened the pull request:
Preliminary Boost networking code that lives inside `net/`.

### Features
- Server info controlled by config file, with bool `IsServer` and string `ServerIp`
- Separate thread to periodically attempt connecting to server
- Client can still start the program without server present

While I think this networking code is not yet perfect/refined, I'd like to get a basic (and somewhat stable) version of it in master before it gets too messy.

richard1 committed:
It's nasty but it works

richard1 committed:
It's nasty but it works

richard1 committed:
It's nasty but it works

richard1 committed:
It's nasty but it works

richard1 committed:
Unnecessary preprocessor def

richard1 committed:
WIN32 preprocessor definition

richard1 committed:
Server changes from GridNetwork branch

richard1 committed:
Add post-event for filesystem dll

richard1 committed:
Read network info from config, periodic connect for client

richard1 committed:
Connection attempts in separate thread

richard1 committed:
Fix merge remnants

richard1 committed:
Tab to space

richard1 committed:
Tab cleanup

richard1 committed:
One more untabify

metakirby5 commented on src/aldente.cpp:
```diff
@@ -256,6 +266,24 @@ void Aldente::go()
+
+    // Try connecting to the host every 5 seconds.
+    while (!connected_to_server) {
+        if (first_attempt || glfwGetTime() - time_last_connect_attempt > 5.f) {
```
Make `time_last_connect_attempt` static to this function if unused elsewhere

metakirby5 commented on src/aldente.cpp:
```diff
@@ -199,6 +199,16 @@ void Aldente::go()
+	client = new NetworkClient(server_host);
+
+	// Repeatedly try establishing connection in separate thread.
+	std::thread(&Aldente::connect_to_server, this).detach();
```
Check for server_host before attempting to connect

metakirby5 commented on src/aldente.cpp:
```diff
@@ -199,6 +199,16 @@ void Aldente::go()
 	tmodel->setScene(scene);
 	scene->root->add_child(tmodel);
+
+	Config::config->get_value(Config::str_is_server, is_server);
```
Update config / config template please, since you added new config keys.

richard1 replied:
🆗

metakirby5 commented on src/net/NetworkClient.h:
```diff
@@ -0,0 +1,60 @@
+#include <thread>
+#include <array>
+
+#define PORT "9000"
```
Make this a config variable then take it in through the constructor?

metakirby5 commented on src/net/NetworkClient.cpp:
```diff
@@ -0,0 +1,73 @@
+    service_thread.join();
+}
+
+bool NetworkClient::init() {
```
Maybe we can call this `connect`? It may also be better to have the API be `connect(host, port)`, and to also have a `disconnect` method if it's easy to implement.

metakirby5 commented on src/net/NetworkClient.cpp:
```diff
@@ -0,0 +1,73 @@
+    return true;
+}
+
+bool NetworkClient::is_initialized() const {
```
Call this `is_connected`?

metakirby5 commented on src/net/NetworkClient.h:
```diff
@@ -0,0 +1,60 @@
+  // Service thread for receiving messages.
+  void run_service();
+
+  boost::asio::io_service io_service;
```
Pass this in as a parameter, since `io_service` can be used for a variety of async boost things.

metakirby5 commented on src/net/NetworkClient.cpp:
```diff
@@ -0,0 +1,73 @@
+    return !message_queue.empty();
+}
+
+std::string NetworkClient::pop_message() {
```
EDIT: On second thought, never mind. `std::string` is probably okay.

metakirby5 commented on src/net/NetworkClient.cpp:
```diff
@@ -0,0 +1,73 @@
+    try {
+        boost::asio::connect(socket, endpoint_iterator);
+    }
+    catch (...) {
```
You some kind of space-waster? Put this on the same line as `}` or I quit this team 😡

metakirby5 commented on src/net/NetworkClient.cpp:
```diff
@@ -0,0 +1,73 @@
+    return initialized;
+}
+
+void NetworkClient::send(std::string message) {
```
Use `std::string &`.

metakirby5 commented on src/net/NetworkClient.cpp:
```diff
@@ -0,0 +1,73 @@
+void NetworkClient::handle_receive(const boost::system::error_code& error, std::size_t bytes_transferred) {
+    if (!error) {
+        std::string message(recv_buffer.data(), recv_buffer.data() + bytes_transferred);
+        message_queue.push(message);
```
EDIT: Ignore most of my comments about synchronous receive; I forgot sockets doing I/O is something that we may have to wait on. I blame lack of sleep.

Consider using a lambda for brevity, though.

richard1 replied:
Ehh I think this function's a little too fat for a lambda... in this case I think it's more readable as a separate function.

metakirby5 commented on src/net/NetworkClient.cpp:
```diff
@@ -0,0 +1,73 @@
+
+void NetworkClient::handle_receive(const boost::system::error_code& error, std::size_t bytes_transferred) {
+    if (!error) {
+        std::string message(recv_buffer.data(), recv_buffer.data() + bytes_transferred);
```
Looks like you're passing in `(&bufchars, &bufchars + n_bytes)` into the `string` constructor. I couldn't find a constructor that matched passing in two pointers, so could you explain what you're doing here?

richard1 replied:
It seems to be constructing a substring and making sure we only keep n_bytes in case of read error or something.  I got this code from somewhere™ and it seems OK so I kept it.

jytang committed:
added libraries needed for net to Makefile

metakirby5 commented on src/net/NetworkClient.cpp:
```diff
@@ -0,0 +1,73 @@
+
+std::string NetworkClient::pop_message() {
+    if (!has_messages())
+        throw std::logic_error("No messages");
```
I'm not a huge fan of exceptions in C++, since they may introduce memory leaks (no garbage collection). Better to return a status boolean and take in a reference parameter to fill with data IMO.

metakirby5 commented on src/net/NetworkClient.cpp:
```diff
@@ -0,0 +1,73 @@
+    boost::asio::write(socket, boost::asio::buffer(message));
+}
+
+bool NetworkClient::has_messages() {
```
Will be obsolete if we use a status boolean return in `pop_message`.

jytang commented:
Well it compiles, certainly! Though do you think we should move the changes made to aldente.{h,cpp} elsewhere? Seems like we'd have to do it anyways in the course of our -ever ongoing refactoring mission intentions-.

metakirby5 commented on src/net/NetworkClient.cpp:
```diff
@@ -0,0 +1,73 @@
+    start_receive();
+}
+
+void NetworkClient::run_service() {
```
EDIT: See above RE: async.

However, if we pass in `io_service`, we can get rid of this and just call `start_recieve` in `init`/`connect`.

jytang reviewed (changes requested):
Nice, but please move changes made to aldente.{cpp,h} into like a \*net setup initialization class thing\*. Or we will have to refactor (with force) in the future.

metakirby5 reviewed (changes requested):
Review part 1 (just the client); part 2 coming soon™

metakirby5 commented on src/net/NetworkClient.cpp:
```diff
@@ -0,0 +1,73 @@
+            io_service.run();
+        }
+        catch (const std::exception& e) {
+            std::cerr << "Client network exception" << e.what();
```
TODO: logging framework

metakirby5 commented on src/net/NetworkServer.cpp:
```diff
@@ -0,0 +1,149 @@
+#include "NetworkServer.h"
+
+TcpConnection::pointer TcpConnection::create(boost::asio::io_service& io_service) {
+  return pointer(new TcpConnection(io_service));
```
I smell memory leak... Any reason why we can't just have a normal constructor and let the user handle allocation?

metakirby5 commented on src/net/NetworkServer.cpp:
```diff
@@ -0,0 +1,149 @@
+  if ((error == boost::asio::error::eof) ||
+      (error == boost::asio::error::connection_reset) ||
+      (error == boost::asio::error::broken_pipe)) {
+    std::cerr << "[send] client disconnected.\n";
```
TODO: logging framework

metakirby5 commented on src/net/NetworkServer.cpp:
```diff
@@ -0,0 +1,149 @@
+    const boost::system::error_code& error) {
+  if (!error) {
+    std::cerr << "Accepted new connection." << std::endl;
+    client_list[++next_id] = new_connection;
```
Not thread-safe, wrap with mutex. Ditto for all other ops with client_list.

metakirby5 commented:
Also: please add `-lboost_thread` to Makefile libs.

jytang commented:
@metakirby5 Already added in 47fb371

metakirby5 commented on src/net/NetworkClient.cpp:
```diff
@@ -0,0 +1,73 @@
+            boost::asio::placeholders::error, boost::asio::placeholders::bytes_transferred));
+}
+
+void NetworkClient::handle_receive(const boost::system::error_code& error, std::size_t bytes_transferred) {
```
Consider just inlining this as a lambda in the `boost::thread` call, since it's quite short.

richard1 replied:
Eh, same reason as above, I like this better as as a separate function.

metakirby5 commented on src/net/NetworkServer.cpp:
```diff
@@ -0,0 +1,149 @@
+  return !message_queue.empty();
+}
+
+std::string TcpConnection::pop_message() {
```
Use a status boolean and output parameter?

metakirby5 commented on src/net/NetworkServer.cpp:
```diff
@@ -0,0 +1,149 @@
+  return true;
+}
+
+bool TcpConnection::has_messages() {
```
Obsolete w/ status boolean?

metakirby5 commented on src/net/NetworkServer.cpp:
```diff
@@ -0,0 +1,149 @@
+TcpConnection::TcpConnection(boost::asio::io_service& io_service)
+  : socket(io_service) {}
+
+void TcpConnection::handle_read(const boost::system::error_code& error, size_t bytes_transferred) {    
```
Inline as lambda?

metakirby5 commented on src/net/NetworkServer.cpp:
```diff
@@ -0,0 +1,149 @@
+  }
+
+  // Wait for and read the next message.
+  socket.async_read_some(boost::asio::buffer(rcvbuf),
```
Just call `start` again here.

metakirby5 commented on src/net/NetworkServer.cpp:
```diff
@@ -0,0 +1,149 @@
+  return message_queue.pop(); 
+}
+
+TcpConnection::TcpConnection(boost::asio::io_service& io_service)
```
Why is this constructor just in the middle of all the other definitions :angery:

metakirby5 commented on src/net/NetworkServer.h:
```diff
@@ -0,0 +1,80 @@
+  static void run(boost::asio::io_service& io_service);
+  
+  boost::asio::io_service io_service;
+  tcp::acceptor acceptor_;
```
Why underscore convention? Nothing else has an underscore.

metakirby5 commented on src/net/NetworkServer.cpp:
```diff
@@ -0,0 +1,149 @@
+  return socket;
+}
+
+void TcpConnection::start() {
```
Consider renaming to `read_loop` or something, `start` isn't super descriptive.

metakirby5 commented on src/net/NetworkServer.cpp:
```diff
@@ -0,0 +1,149 @@
+  start_accept();
+  
+  // Run the io_service in a separate thread so it's non-blocking.
+  std::thread(TcpServer::run, std::ref(io_service)).detach();
```
Again, pass in the `io_service` as a constructor arg. We probably want to reuse one `io_service` across the entire program, should we need more than one async boost thing.

metakirby5 commented on src/net/NetworkServer.cpp:
```diff
@@ -0,0 +1,149 @@
+
+void TcpServer::send_to_all(std::string message) {
+  // No clients connected.
+  if (client_list.size() == 0) {
```
`.empty()`

metakirby5 commented on src/net/NetworkServer.cpp:
```diff
@@ -0,0 +1,149 @@
+  std::vector<std::string> messages;
+  
+  // No clients connected.
+  if (client_list.size() == 0) {
```
`.empty()`

metakirby5 commented on src/net/NetworkServer.cpp:
```diff
@@ -0,0 +1,149 @@
+      bool success = c->second->send(message);
+      if (!success) {
+          std::cerr << "Write failed!\n";
+          c = client_list.erase(c);
```
If a write fails we just drop the client? Maybe comment noting this.

metakirby5 commented on src/net/NetworkServer.cpp:
```diff
@@ -0,0 +1,149 @@
+  for (auto c = client_list.begin(); c != client_list.end(); /* empty */) {
+      bool success = c->second->send(message);
+      if (!success) {
+          std::cerr << "Write failed!\n";
```
TODO: logging framework

metakirby5 commented on src/net/NetworkServer.cpp:
```diff
@@ -0,0 +1,149 @@
+    message_queue.push(rcvbuf);
+  }
+  else if (error != boost::asio::error::eof) {
+    std::cerr << "FATAL handle_read error: " << error << "\n";
```
TODO: logging framework

metakirby5 commented on src/net/NetworkServer.cpp:
```diff
@@ -0,0 +1,149 @@
+void TcpServer::handle_accept(TcpConnection::pointer new_connection,
+    const boost::system::error_code& error) {
+  if (!error) {
+    std::cerr << "Accepted new connection." << std::endl;
```
TODO: logging framework

metakirby5 commented on src/net/NetworkServer.cpp:
```diff
@@ -0,0 +1,149 @@
+  for (auto const &c : client_list) {
+    while (c.second->has_messages()) {
+      auto message = c.second->pop_message();
+      messages.push_back(message);
```
We probably want some indication of which client the message came from. Maybe use `vector<pair<client_id, message>>`? Or even a map of `client_id -> vector<message>`.

metakirby5 commented on src/net/NetworkServer.cpp:
```diff
@@ -0,0 +1,149 @@
+    TcpConnection::create(acceptor_.get_io_service());
+
+  acceptor_.async_accept(new_connection->get_socket(),
+      boost::bind(&TcpServer::handle_accept, this, new_connection,
```
Inline with lambda?

metakirby5 commented on src/net/NetworkServer.cpp:
```diff
@@ -0,0 +1,149 @@
+  start_accept();
+}
+
+void TcpServer::run(boost::asio::io_service& io_service) {
```
Pass in `io_service` to constructor and start externally.

metakirby5 commented on src/net/NetworkServer.h:
```diff
@@ -0,0 +1,80 @@
+
+  tcp::socket socket;
+  char rcvbuf[BUFSIZ];
+  ThreadSafeQueue<std::string> message_queue;
```
Reminder: http://www.boost.org/doc/libs/1_53_0/doc/html/boost/lockfree/queue.html

richard1 replied:
Sadly the the lockfree queue data types must have a "trivial copy assignment and destructor" which std::string doesn't have.

http://stackoverflow.com/questions/17452491/how-to-put-stdstring-into-boostlockfreequeue-or-alternative

metakirby5 reviewed (changes requested):
Second part of the review. Client and TCPConnection seem to share a lot of logic, so consider linking those classes? Not necessary if you think they're distinct enough though.

richard1 committed:
Move aldente net stuff to NetworkManager

richard1 committed:
Most PR changes for client

richard1 committed:
Revert changes in aldente.cpp

richard1 committed:
Address most server-related comments

richard1 commented:
I've addressed most comments except for the common io_service; will continue working on that.

Other changes:
- Rename `TcpConnection` to `Connection`, `TcpServer` to `NetworkServer`
- Pull out net code from aldente.{cpp, h} and put into new `NetworkManager`

richard1 committed:
Single io_service, fix indentation

richard1 committed:
Use a tab for consistency

richard1 commented:
Addressed PR comments and refactored a bit.  Should be ready for another review.

Also @jytang I left some network code in aldente.cpp for ease of testing; I can remove them after code review and before merging if you want.

jytang committed:
removed extra qualifications in network headers

metakirby5 committed:
start_receive lambda

metakirby5 commented on src/NetworkManager.cpp:
```diff
@@ -0,0 +1,63 @@
+    }
+}
+
+void NetworkManager::attempt_connection() {
```
This logic should be moved into `NetworkClient`.

richard1 replied:
If we're keeping `NetworkManager`, I'd like to leave this in here for now.  We're defining the reconnection strategy at a higher level than the actual client, which I like.  I can see this moving into `NetworkClient` in the future though.

metakirby5 commented on src/NetworkManager.cpp:
```diff
@@ -0,0 +1,63 @@
+    Config::config->get_value(Config::str_server_ip, server_host);
+    Config::config->get_value(Config::str_port, port);
+
+    server = is_server ? new NetworkServer(&io_service, port) : nullptr;
```
We C++ now, pass in by reference my man.

metakirby5 replied:
Also, any reason for the `new`? We could just allocate it on the stack.

richard1 replied:
I'd like to have new since I have server and client objects as class variables, so their initial values are null pointers.  For the `is_server = false` case, server would be null instead of an empty `Server` object.

metakirby5 commented on src/NetworkManager.cpp:
```diff
@@ -0,0 +1,63 @@
+#include "NetworkManager.h"
```
I feel like the `NetworkManager` class is overall unnecessary if we keep reconnect functionality in `NetworkClient`. We can keep it around for now for reference though.

metakirby5 commented on src/NetworkManager.cpp:
```diff
@@ -0,0 +1,63 @@
+    while (!is_connected) {
+        if (first_attempt || glfwGetTime() - time_last_connect_attempt > CONN_RETRY_SEC) {
+            first_attempt = false;
+            std::cerr << "Attempting to connect to " << server_host << "...\n";
```
MUH LOGGING

richard1 replied:
Boost logging library is a pain to link.  Gonna look into it later in a later PR.

metakirby5 replied:
:angery:

metakirby5 commented on src/NetworkManager.h:
```diff
@@ -0,0 +1,37 @@
+    bool is_connected = false;
+    NetworkServer* server;
+    NetworkClient* client;
+    boost::asio::io_service io_service;
```
Again, `io_service` is not network-specific, so I'm opposed to putting it in the `NetworkManager`.

richard1 replied:
Hmm, we're not exactly using it for anything else, so I'm okay with putting it here.  It helps hide this detail from higher levels.

metakirby5 commented on src/net/NetworkClient.cpp:
```diff
@@ -0,0 +1,54 @@
+#include "NetworkClient.h"
+
+NetworkClient::NetworkClient(boost::asio::io_service* ios) :
```
Pls use `const boost::asio::io_service &ios`

richard1 replied:
Fails to compile with `const`

metakirby5 commented on src/net/NetworkClient.cpp:
```diff
@@ -0,0 +1,54 @@
+    return true;
+}
+
+bool NetworkClient::read_message(std::string& message) {
```
This is actually an instance where it is better to not use a reference - by making it a pointer, it's more clear that it is an input parameter.

metakirby5 commented on src/net/NetworkClient.cpp:
```diff
@@ -0,0 +1,54 @@
+    return connected;
+}
+
+bool NetworkClient::send(std::string& message) {
```
`const string &message`

metakirby5 commented on src/net/NetworkServer.cpp:
```diff
@@ -0,0 +1,124 @@
+    return socket;
+}
+
+void Connection::read_loop() {
```
LAMBDA IT UP

metakirby5 commented on src/net/NetworkServer.cpp:
```diff
@@ -0,0 +1,124 @@
+    if ((error == boost::asio::error::eof) ||
+        (error == boost::asio::error::connection_reset) ||
+        (error == boost::asio::error::broken_pipe)) {
+        std::cerr << "[send] client disconnected.\n";
```
MUH LOGS

metakirby5 commented on src/net/NetworkServer.cpp:
```diff
@@ -0,0 +1,124 @@
+    return true;
+}
+
+bool Connection::read_message(std::string& message) {
```
Use pointer for same reason as above. `string *message`

metakirby5 commented on src/net/NetworkServer.cpp:
```diff
@@ -0,0 +1,124 @@
+            boost::asio::placeholders::bytes_transferred));
+}
+
+bool Connection::send(std::string message) {
```
`const string &message`

metakirby5 commented on src/net/NetworkServer.cpp:
```diff
@@ -0,0 +1,124 @@
+#include "NetworkServer.h"
```
Code duplicated between `Connection` and `NetworkClient`. Put a `Connection` inside `NetworkClient`? (And move `Connection` to its own file.)

metakirby5 commented on src/net/NetworkServer.cpp:
```diff
@@ -0,0 +1,124 @@
+void Connection::handle_read(const boost::system::error_code& error, size_t bytes_transferred) {
+    if (!error) {
+        message_queue.push(rcvbuf);
+        std::cerr << "[s] recv: " << std::string(rcvbuf);
```
LOGS

metakirby5 commented on src/net/NetworkClient.cpp:
```diff
@@ -0,0 +1,54 @@
+    io_service(ios), socket(*ios), connected(false) {
+}
+
+bool NetworkClient::connect(std::string& host, unsigned int port) {
```
`const string &host`

metakirby5 commented on src/net/NetworkServer.cpp:
```diff
@@ -0,0 +1,124 @@
+    read_loop();
+}
+
+NetworkServer::NetworkServer(boost::asio::io_service* ios, unsigned int port) :
```
`const boost::asio::io_service &ios`

metakirby5 commented on src/net/NetworkServer.cpp:
```diff
@@ -0,0 +1,124 @@
+    for (auto c = client_list.begin(); c != client_list.end(); /* empty */) {
+        bool success = c->second->send(message);
+        if (!success) {
+            std::cerr << "Write failed!\n";
```
LOGS

metakirby5 commented on src/net/NetworkServer.cpp:
```diff
@@ -0,0 +1,124 @@
+}
+
+std::map<int, std::vector<std::string>> NetworkServer::read_all_messages() {
+    std::map<int, std::vector<std::string>> messages;
```
`std::unordered_map` if order does not matter when returning

richard1 committed:
std::string -> string

metakirby5 commented on src/net/NetworkServer.cpp:
```diff
@@ -0,0 +1,124 @@
+    Connection::pointer new_connection =
+        Connection::pointer(new Connection(acceptor.get_io_service()));
+
+    acceptor.async_accept(new_connection->get_socket(),
```
LAMBDA IT UP

metakirby5 commented on src/net/NetworkServer.cpp:
```diff
@@ -0,0 +1,124 @@
+void NetworkServer::handle_accept(Connection::pointer new_connection,
+    const boost::system::error_code& error) {
+    if (!error) {
+        std::cerr << "Accepted new connection." << std::endl;
```
LOGS

metakirby5 commented on src/net/NetworkServer.cpp:
```diff
@@ -0,0 +1,124 @@
+        std::cerr << "Accepted new connection." << std::endl;
+        unique_lock<mutex> lock(client_list_mutex);
+        client_list[++next_id] = new_connection;
+        lock.unlock();
```
Unnecessary, as it will be unlocked once it goes out of scope (I think).

metakirby5 commented on src/net/NetworkServer.cpp:
```diff
@@ -0,0 +1,124 @@
+        unique_lock<mutex> lock(client_list_mutex);
+        client_list[++next_id] = new_connection;
+        lock.unlock();
+        new_connection->read_loop();
```
Consider renaming to `start_async_read_loop`, since my initial impression is that calling this method will block the thread until the loop is done.

metakirby5 commented on src/net/NetworkServer.cpp:
```diff
@@ -0,0 +1,124 @@
+}
+
+void NetworkServer::start_accept() {
+    Connection::pointer new_connection =
```
What's the reason for using a shared pointer here? Why not stack?

Also, I advise against typedefing pointer, since it was a bit confusing for me.

richard1 replied:
Apparently shared pointers help maintain exception safety better than normal pointers, and so lots of asio stuff uses shared pointers: http://stackoverflow.com/a/4099375/2014825

metakirby5 reviewed (changes requested):
this code is 🔥

richard1 committed:
Part 2 of PR comments

richard1 committed:
Remove and reuse connection class

richard1 commented:
Ready for part 3 of the code review 💯

metakirby5 commented on src/net/Connection.cpp:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    boost::system::error_code error;
+    boost::asio::write(socket, boost::asio::buffer(message), error);
```
Since we're doing async read, might as well do async write with lambda? http://www.boost.org/doc/libs/1_41_0/doc/html/boost_asio/reference/async_write/overload1.html

metakirby5 reviewed (changes requested):
SO CLEAN I COULD BATHE IN IT

:ship: it :boy: s

richard1 committed:
Add comment about async write

richard1 committed:
Reset changes to aldente.cpp to reduce conflicts later
